tests/patcher_test.py

```diff
@@ -294,6 +294,58 @@ def test_subprocess_after_monkey_patch():
     assert output.rstrip() == b'pass'
 
 
+def test_fork_after_monkey_patch():
+    code = '''\
+import eventlet
+import eventlet.patcher
+eventlet.monkey_patch()
+import os
+import sys
+import threading
+_threading = eventlet.patcher.original('threading')
+import eventlet.green.threading
+def target():
+    eventlet.sleep(0.1)
+def check(n, mod, tag):
+    assert len(mod._active) == n, 'Expected {} {} threads, got {}'.format(n, tag, mod._active)
+threads = [
+    threading.Thread(target=target, name='patched'),
+    _threading.Thread(target=target, name='original-1'),
+    _threading.Thread(target=target, name='original-2'),
+    eventlet.green.threading.Thread(target=target, name='green-1'),
+    eventlet.green.threading.Thread(target=target, name='green-2'),
+    eventlet.green.threading.Thread(target=target, name='green-3'),
+]
+for t in threads: t.start()
+check(2, threading, 'pre-fork patched')
+check(3, _threading, 'pre-fork original')
+check(4, eventlet.green.threading, 'pre-fork green')
+if os.fork() == 0:
+    # Inside the child, we should only have a main thread,
+    # but old pythons make it difficult to ensure
+    if sys.version_info >= (3, 7):
+        check(1, threading, 'child post-fork patched')
+        check(1, _threading, 'child post-fork original')
```

Comment:
so this wasn't expected - and it may be the crux of the new 3.7 feature that's applying some cleaning.

This is closer to what I expected from my experience with python and forking with threads:

```
diff --git a/tests/patcher_test.py b/tests/patcher_test.py
index 0bebe39..ff8dc24 100644
--- a/tests/patcher_test.py
+++ b/tests/patcher_test.py
@@ -326,6 +326,9 @@ if os.fork() == 0:
     if sys.version_info >= (3, 7):
         check(1, threading, 'child post-fork patched')
         check(1, _threading, 'child post-fork original')
+    else:
+        check(2, threading, 'child post-fork patched')
+        check(3, _threading, 'child post-fork original')
     check(1, eventlet.green.threading, 'child post-fork green')
     exit()
 else:
```

But I don't have enough experience outside of python to say if that's "reasonable"

Reply:
IDK -- seems pretty consistent with how (unpatched) stdlib threading works across a variety of versions:
```
$ cat test_threads_and_forking.py 
import os
import threading
import time

threads = [threading.Thread(target=time.sleep, args=(1,)) for _ in range(2)]
for t in threads: t.start()
time.sleep(0.1)  # Give py2 threads a chance to actually start
print('pre-fork: {}'.format(len(threading._active)))
label = 'child' if os.fork() == 0 else 'parent'
print('post-fork ({}): {}'.format(label, len(threading._active)))
print('liveness ({}): {}'.format(label, [t.is_alive() for t in threads]))
for t in threads: t.join()
if label == 'child':
	exit()
else:
	os.wait()
$ for p in python2.7 python3.5 python3.6 python3.7 python3.8; do $p --version ; $p test_threads_and_forking.py ; echo ; done
Python 2.7.17
pre-fork: 3
post-fork (parent): 3
liveness (parent): [True, True]
post-fork (child): 1
liveness (child): [False, False]

Python 3.5.9
pre-fork: 3
post-fork (parent): 3
liveness (parent): [True, True]
post-fork (child): 1
liveness (child): [False, False]

Python 3.6.9
pre-fork: 3
post-fork (parent): 3
liveness (parent): [True, True]
post-fork (child): 1
liveness (child): [False, False]

Python 3.7.7
pre-fork: 3
post-fork (parent): 3
liveness (parent): [True, True]
post-fork (child): 1
liveness (child): [False, False]

Python 3.8.3
pre-fork: 3
post-fork (parent): 3
liveness (parent): [True, True]
post-fork (child): 1
liveness (child): [False, False]
```